### ISSUE TITLE
Add a zero cost assertion mechanism

### ISSUE THREAD

**Issue by minborg** [OpenHFT/Chronicle-Core#221]
Normal assertions using the `assert` keyword have several drawbacks:

* It requires byte code space, thereby increasing code space
* Interpretation mode is slower because more code has to be interpreted
* Compile mode might be substantially slower because the inlining threshold will be consumed by assertion code
* It always produces the same kind of Error (i.e. AssertionError)
* Unless accompanied with an assert message, it provides little extra information as to which invariant did not hold (e.g. "" vs. "The provided index was non-negative (=-3))

It is possible to write an equivalent assertion mechanism that solves all the above (i.e. with no impact on byte code size, interpreting, and compiled code performance). The drawback is that switching on assertions will be non-standard.

Having an assertion mechanism in place is key to discovering bugs in our libraries.

**Comment by minborg** [comment]
So, a class that looks like this:

```
public final class EnableTest2 {

    public static void test() {
        assert AssertUtil.SKIP_ASSERTIONS || IntAssert.assertZero(3);
    }

}
```

Will have the following byte code with `SKIP_ASSERTIONS = false`:


```
% javap -c EnableTest 
Warning: Binary file EnableTest contains net.openhft.chronicle.core.invariant.EnableTest
Compiled from "EnableTest.java"
public final class net.openhft.chronicle.core.invariant.EnableTest {
  public net.openhft.chronicle.core.invariant.EnableTest();
    Code:
       0: aload_0
       1: invokespecial #1                  // Method java/lang/Object."<init>":()V
       4: return

  public static void test();
    Code:
       0: iconst_3
       1: invokestatic  #2                  // Method net/openhft/chronicle/core/invariant/IntAssert.assertZero:(I)V
       4: return
}
```

And with `SKIP_ASSERTIONS = true`:

```

% javap -c  target/test-classes/net/openhft/chronicle/core/invariant/EnableTest2.class
Compiled from "EnableTest2.java"
public final class net.openhft.chronicle.core.invariant.EnableTest2 {
  public net.openhft.chronicle.core.invariant.EnableTest2();
    Code:
       0: aload_0
       1: invokespecial #1                  // Method java/lang/Object."<init>":()V
       4: return

  public static void test();
    Code:
       0: return
}
```

**Comment by minborg** [comment]
I have added several useful features to the zero-cost assertion mechanism. See the example below:

Here is an original class:

```
import static net.openhft.chronicle.core.util.AssertUtil.SKIP_ASSERTIONS;
import static net.openhft.chronicle.core.util.IntCondition.NON_NEGATIVE;
import static net.openhft.chronicle.core.util.Ints.*;

public final class AssertTest {

    public static void testWithAssertDirectly() {
        assert SKIP_ASSERTIONS || NON_NEGATIVE.test(3);
    }

    public static void testWithAssertDirectlyWithText() {
        assert SKIP_ASSERTIONS || NON_NEGATIVE.test(3) : failDescription(NON_NEGATIVE, 3);
    }

    public static void testWithAssertMethod() {
        assertIfEnabled(NON_NEGATIVE, 3);
    }

}

```

It has the following byte code with `SKIP_ASSERTIONS = false`:

```
% javap -c  target/test-classes/net/openhft/chronicle/core/util/AssertTest.class
Compiled from "AssertTest.java"
public final class net.openhft.chronicle.core.util.AssertTest {
  static final boolean $assertionsDisabled;

  public net.openhft.chronicle.core.util.AssertTest();
    Code:
       0: aload_0
       1: invokespecial #2                  // Method java/lang/Object."<init>":()V
       4: return

  public static void testWithAssertDirectly();
    Code:
       0: getstatic     #3                  // Field $assertionsDisabled:Z
       3: ifne          24
       6: getstatic     #4                  // Field net/openhft/chronicle/core/util/IntCondition.NON_NEGATIVE:Lnet/openhft/chronicle/core/util/IntCondition;
       9: iconst_3
      10: invokevirtual #5                  // Method net/openhft/chronicle/core/util/IntCondition.test:(I)Z
      13: ifne          24
      16: new           #6                  // class java/lang/AssertionError
      19: dup
      20: invokespecial #7                  // Method java/lang/AssertionError."<init>":()V
      23: athrow
      24: return

  public static void testWithAssertDirectlyWithText();
    Code:
       0: getstatic     #3                  // Field $assertionsDisabled:Z
       3: ifne          31
       6: getstatic     #4                  // Field net/openhft/chronicle/core/util/IntCondition.NON_NEGATIVE:Lnet/openhft/chronicle/core/util/IntCondition;
       9: iconst_3
      10: invokevirtual #5                  // Method net/openhft/chronicle/core/util/IntCondition.test:(I)Z
      13: ifne          31
      16: new           #6                  // class java/lang/AssertionError
      19: dup
      20: getstatic     #4                  // Field net/openhft/chronicle/core/util/IntCondition.NON_NEGATIVE:Lnet/openhft/chronicle/core/util/IntCondition;
      23: iconst_3
      24: invokestatic  #8                  // Method net/openhft/chronicle/core/util/Ints.failDescription:(Ljava/util/function/IntPredicate;I)Ljava/lang/String;
      27: invokespecial #9                  // Method java/lang/AssertionError."<init>":(Ljava/lang/Object;)V
      30: athrow
      31: return

  public static void testWithAssertMethod();
    Code:
       0: getstatic     #4                  // Field net/openhft/chronicle/core/util/IntCondition.NON_NEGATIVE:Lnet/openhft/chronicle/core/util/IntCondition;
       3: iconst_3
       4: invokestatic  #10                 // Method net/openhft/chronicle/core/util/Ints.assertIfEnabled:(Ljava/util/function/IntPredicate;I)V
       7: return

  static {};
    Code:
       0: ldc           #11                 // class net/openhft/chronicle/core/util/AssertTest
       2: invokevirtual #12                 // Method java/lang/Class.desiredAssertionStatus:()Z
       5: ifne          12
       8: iconst_1
       9: goto          13
      12: iconst_0
      13: putstatic     #3                  // Field $assertionsDisabled:Z
      16: return
}
```

But it has the following byte code with `SKIP_ASSERTIONS = true`:

```
javap -c  target/test-classes/net/openhft/chronicle/core/util/AssertTest.class      
Compiled from "AssertTest.java"
public final class net.openhft.chronicle.core.util.AssertTest {
  public net.openhft.chronicle.core.util.AssertTest();
    Code:
       0: aload_0
       1: invokespecial #2                  // Method java/lang/Object."<init>":()V
       4: return

  public static void testWithAssertDirectly();
    Code:
       0: return

  public static void testWithAssertDirectlyWithText();
    Code:
       0: return

  public static void testWithAssertMethod();
    Code:
       0: getstatic     #3                  // Field net/openhft/chronicle/core/util/IntCondition.NON_NEGATIVE:Lnet/openhft/chronicle/core/util/IntCondition;
       3: iconst_3
       4: invokestatic  #4                  // Method net/openhft/chronicle/core/util/Ints.assertIfEnabled:(Ljava/util/function/IntPredicate;I)V
       7: return
}

```
As can be seen, the cost is zero when `SKIP_ASSERTIONS = true` in the first two cases.

Performance critical code should use one of the first two schemes devised above to assert invariants. The third, more convenient form, can be used for non-performance critical code.

**Comment by hft-team-city** [comment]
Released in [Chronicle-Core-2.21ea47](https://github.com/OpenHFT/Chronicle-Core/releases/chronicle-core-2.21ea47), [BOM-2.21ea119](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.21ea119)
